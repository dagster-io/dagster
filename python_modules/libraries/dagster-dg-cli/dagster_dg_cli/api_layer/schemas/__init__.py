"""Pydantic schemas for Dagster Plus API."""
