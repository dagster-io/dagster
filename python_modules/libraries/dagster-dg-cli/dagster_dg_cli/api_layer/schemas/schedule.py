"""Schedule models for REST-like API."""

from typing import Optional

from pydantic import BaseModel


class DgApiSchedule(BaseModel):
    """Schedule resource model."""

    id: str
    name: str
    cron_schedule: str
    pipeline_name: str  # Note: will rename to job_name in output formatting per INTERNAL_GRAPHQL_USAGE.md guidance
    description: Optional[str]
    execution_timezone: str
    tags: list[dict]
    metadata_entries: list[dict]

    class Config:
        from_attributes = True


class DgApiScheduleList(BaseModel):
    """GET /api/schedules response."""

    items: list[DgApiSchedule]
