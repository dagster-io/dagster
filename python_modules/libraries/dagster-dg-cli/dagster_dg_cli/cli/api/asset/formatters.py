"""Asset output formatters for CLI display."""

import re
from typing import TYPE_CHECKING, Literal

from rich.console import Console
from rich.table import Table
from rich.text import Text

from dagster_dg_cli.cli.api.asset.display_models import (
    AssetDisplay,
    AssetListDisplay,
    MetadataSection,
    asset_list_to_display,
    asset_to_display,
)

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.asset import DgApiAsset, DgApiAssetList

OutputFormat = Literal["table", "json", "markdown"]


def _enhance_sql_formatting(sql_content: str) -> str:
    """Apply consistent indentation to SQL content within code blocks."""
    lines = sql_content.split("\n")
    formatted_lines = []

    for line in lines:
        stripped = line.strip()
        if stripped:
            if any(
                stripped.lower().startswith(keyword)
                for keyword in [
                    "select",
                    "from",
                    "where",
                    "with",
                    "left join",
                    "right join",
                    "inner join",
                    "join",
                ]
            ):
                formatted_lines.append(f"    {stripped}")
            elif stripped.startswith(("),", ") as", ")", "and ", "or ", "on ")):
                formatted_lines.append(f"    {stripped}")
            else:
                formatted_lines.append(f"      {stripped}")
        else:
            formatted_lines.append("")

    return "\n".join(formatted_lines)


def _process_description_with_code_blocks(description: str, format_sql: bool = True) -> list[Text]:
    """Process description text, enhancing SQL code blocks while preserving structure."""
    if not description:
        return []

    lines = []
    current_text = []
    in_code_block = False
    code_block_type = None
    code_content = []

    for line in description.split("\n"):
        code_block_match = re.match(r"```(\w+)?", line.strip())
        if code_block_match and not in_code_block:
            if current_text:
                for text_line in current_text:
                    lines.append(Text(f"  {text_line}" if text_line.strip() else ""))
                current_text = []

            in_code_block = True
            code_block_type = code_block_match.group(1) or ""
            lines.append(Text(f"  {line.strip()}"))
            code_content = []

        elif line.strip() == "```" and in_code_block:
            if code_content and code_block_type and code_block_type.lower() == "sql" and format_sql:
                enhanced_sql = _enhance_sql_formatting("\n".join(code_content))
                for sql_line in enhanced_sql.split("\n"):
                    lines.append(Text(f"  {sql_line}" if sql_line.strip() else ""))
            else:
                for code_line in code_content:
                    lines.append(Text(f"  {code_line}"))

            lines.append(Text(f"  {line.strip()}"))
            in_code_block = False
            code_block_type = None
            code_content = []

        elif in_code_block:
            code_content.append(line)

        else:
            current_text.append(line)

    if current_text:
        for text_line in current_text:
            lines.append(Text(f"  {text_line}" if text_line.strip() else ""))

    return lines


def _create_horizontal_separator(section_name: str, section_color: str, width: int = 70) -> Text:
    """Create a styled horizontal separator with section name."""
    prefix = "━━━ "
    suffix_length = width - len(prefix) - len(section_name) - 1
    suffix = "━" * max(suffix_length, 3)

    separator_text = Text(f"{prefix}{section_name} {suffix}", style=section_color)
    return separator_text


def format_assets(assets: "DgApiAssetList", output_format: OutputFormat = "table") -> str:
    """Format asset list for output."""
    if output_format == "json":
        return assets.model_dump_json(indent=2)

    display = asset_list_to_display(assets)

    if output_format == "markdown":
        return _format_asset_list_display_as_markdown(display)

    return _format_asset_list_display_as_table(display)


def format_asset(asset: "DgApiAsset", output_format: OutputFormat = "table") -> str:
    """Format single asset for output."""
    if output_format == "json":
        return asset.model_dump_json(indent=2)

    display = asset_to_display(asset)

    if output_format == "markdown":
        return _format_asset_display_as_markdown(display)

    return _format_asset_display_as_table(display)


def _format_asset_list_display_as_table(display: AssetListDisplay) -> str:
    """Format asset list display model as table."""
    if not display.items:
        return "No assets found."

    table = Table(border_style="dim")
    table.add_column("Asset Key", style="bold cyan", no_wrap=True, min_width=20)
    table.add_column("Group", style="bold", min_width=12)
    table.add_column("Kinds", style="bold", min_width=12)
    table.add_column("Description", style="bold", max_width=60, no_wrap=True)

    for item in display.items:
        table.add_row(
            item.asset_key,
            item.group_name,
            item.kinds_text,
            item.description_text,
        )

    console = Console()
    with console.capture() as capture:
        console.print(table)
    return capture.get().rstrip()


def _format_asset_list_display_as_markdown(display: AssetListDisplay) -> str:
    """Format asset list display model as markdown."""
    if not display.items:
        return "No assets found."

    lines = [
        "# Assets",
        "",
        "| Asset Key | Group | Kinds | Description |",
        "|-----------|-------|-------|-------------|",
    ]

    for item in display.items:
        description_text = item.description_text.replace("|", "\\|").replace("\n", " ")
        if len(description_text) > 100:
            description_text = description_text[:97] + "..."

        lines.append(
            f"| {item.asset_key} | {item.group_name} | {item.kinds_text} | {description_text} |"
        )

    return "\n".join(lines)


def _format_asset_display_as_table(display: AssetDisplay) -> str:
    """Format asset display model as rich table."""
    console = Console()
    lines = []

    header_text = Text(f"Asset: {display.asset_key}", style="bold cyan")
    if display.group_name:
        header_text.append(f" [{display.group_name}]", style="dim")
    lines.append(header_text)

    if display.kinds:
        kinds_text = Text("Kinds: ", style="bold")
        kinds_text.append(", ".join(display.kinds), style="blue")
        lines.append(kinds_text)

    if display.tags:
        tags_text = Text("Tags: ", style="bold")
        tag_strings = [f"{key}={value}" for key, value in display.tags]
        tags_text.append(", ".join(tag_strings), style="green")
        lines.append(tags_text)

    lines.append(Text(""))

    if display.description:
        lines.append(_create_horizontal_separator("Description", "blue"))
        desc_lines = _process_description_with_code_blocks(display.description, format_sql=True)
        lines.extend(desc_lines)
        lines.append(Text(""))

    for section in display.metadata_sections:
        lines.extend(_format_metadata_section_display(section))

    lines.append(_create_horizontal_separator("Technical Details", "dim"))
    id_text = Text("ID:", style="bold")
    id_text.append(f" {display.id}", style="dim")
    lines.append(id_text)

    with console.capture() as capture:
        for line in lines:
            console.print(line)

    return capture.get().rstrip()


def _format_metadata_section_display(section: MetadataSection) -> list[Text]:
    """Format a metadata section display model."""
    if not section.entries:
        return []

    lines = []
    lines.append(_create_horizontal_separator(section.name, section.color))

    for entry in section.entries:
        if entry.is_url:
            field_text = Text(f"{entry.display_name}:", style="bold")
            field_text.append(f" {entry.value}", style="blue")
            lines.append(field_text)
        else:
            field_text = Text(f"{entry.display_name}:", style="bold")
            field_text.append(f" {entry.value}")
            lines.append(field_text)

        lines.append(Text(f"       ({entry.label})", style="dim"))

    lines.append(Text(""))
    return lines


def _format_asset_display_as_markdown(display: AssetDisplay) -> str:
    """Format asset display model as markdown."""
    lines = [f"# Asset: {display.asset_key}"]

    if display.group_name:
        lines.append(f"**Group:** {display.group_name}")

    if display.kinds:
        lines.append(f"**Kinds:** {', '.join(display.kinds)}")

    lines.append("")

    if display.description:
        lines.extend(["## Description", "", display.description, ""])

    for section in display.metadata_sections:
        if section.name == "Metadata":
            lines.extend(["## Metadata", ""])
        else:
            lines.extend([f"## {section.name} Information", ""])

        for entry in section.entries:
            lines.append(f"- **{entry.display_name}:** {entry.value}")
            lines.append(f"  - _({entry.label})_")
        lines.append("")

    lines.extend(["## Technical Details", "", f"**ID:** {display.id}"])

    return "\n".join(lines)
