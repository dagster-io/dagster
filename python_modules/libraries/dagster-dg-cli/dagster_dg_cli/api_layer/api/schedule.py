"""Schedule endpoints - REST-like interface."""

from dataclasses import dataclass
from typing import TYPE_CHECKING

from dagster_dg_cli.api_layer.graphql_adapter.schedule import (
    get_dg_plus_api_schedule_via_graphql,
    list_dg_plus_api_schedules_via_graphql,
)
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.schedule import DgApiSchedule, DgApiScheduleList


@dataclass(frozen=True)
class DgApiScheduleApi:
    client: IGraphQLClient

    def list_schedules(self) -> "DgApiScheduleList":
        """List schedules."""
        return list_dg_plus_api_schedules_via_graphql(self.client)

    def get_schedule(self, schedule_name: str) -> "DgApiSchedule":
        """Get single schedule by name."""
        return get_dg_plus_api_schedule_via_graphql(self.client, schedule_name)
