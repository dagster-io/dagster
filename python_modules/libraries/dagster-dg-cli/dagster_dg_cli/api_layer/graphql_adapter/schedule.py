"""GraphQL implementation for schedule operations."""

from dagster_dg_cli.api_layer.schemas.schedule import DgApiSchedule, DgApiScheduleList
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

# GraphQL queries
SCHEDULES_QUERY = """
query Schedules {
    workspaceOrError {
        __typename
        ... on Workspace {
            locationEntries {
                __typename
                ... on WorkspaceLocationEntry {
                    name
                    locationOrLoadError {
                        __typename
                        ... on RepositoryLocation {
                            repositories {
                                name
                                schedules {
                                    id
                                    name
                                    cronSchedule
                                    pipelineName
                                    description
                                    executionTimezone
                                    tags {
                                        key
                                        value
                                    }
                                    metadataEntries {
                                        label
                                        description
                                        ... on TextMetadataEntry {
                                            text
                                        }
                                        ... on UrlMetadataEntry {
                                            url
                                        }
                                        ... on PathMetadataEntry {
                                            path
                                        }
                                        ... on JsonMetadataEntry {
                                            jsonString
                                        }
                                        ... on MarkdownMetadataEntry {
                                            mdStr
                                        }
                                        ... on PythonArtifactMetadataEntry {
                                            module
                                            name
                                        }
                                        ... on FloatMetadataEntry {
                                            floatValue
                                        }
                                        ... on IntMetadataEntry {
                                            intValue
                                        }
                                        ... on BoolMetadataEntry {
                                            boolValue
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        ... on PythonError {
            message
        }
    }
}
"""

SCHEDULE_QUERY = """
query Schedule($scheduleName: String!, $repositoryName: String!, $repositoryLocationName: String!) {
    scheduleOrError(
        scheduleSelector: {
            scheduleName: $scheduleName
            repositoryName: $repositoryName
            repositoryLocationName: $repositoryLocationName
        }
    ) {
        __typename
        ... on Schedule {
            id
            name
            cronSchedule
            pipelineName
            description
            executionTimezone
            tags {
                key
                value
            }
            metadataEntries {
                label
                description
                ... on TextMetadataEntry {
                    text
                }
                ... on UrlMetadataEntry {
                    url
                }
                ... on PathMetadataEntry {
                    path
                }
                ... on JsonMetadataEntry {
                    jsonString
                }
                ... on MarkdownMetadataEntry {
                    mdStr
                }
                ... on PythonArtifactMetadataEntry {
                    module
                    name
                }
                ... on FloatMetadataEntry {
                    floatValue
                }
                ... on IntMetadataEntry {
                    intValue
                }
                ... on BoolMetadataEntry {
                    boolValue
                }
            }
        }
        ... on ScheduleNotFoundError {
            message
        }
        ... on PythonError {
            message
        }
    }
}
"""


def _transform_metadata_entries(metadata_entries):
    """Transform GraphQL metadata entries to dict format."""
    transformed = []
    for entry in metadata_entries:
        metadata_dict = {
            "label": entry.get("label", ""),
            "description": entry.get("description", ""),
        }

        # Add type-specific fields
        for field in [
            "text",
            "url",
            "path",
            "jsonString",
            "mdStr",
            "floatValue",
            "intValue",
            "boolValue",
        ]:
            if field in entry:
                metadata_dict[field] = entry[field]
        if "module" in entry and "name" in entry:
            metadata_dict.update({"module": entry["module"], "name": entry["name"]})

        transformed.append(metadata_dict)
    return transformed


def _transform_tags(tags):
    """Transform GraphQL tags to dict format."""
    return [{"key": tag.get("key", ""), "value": tag.get("value", "")} for tag in tags]


def list_dg_plus_api_schedules_via_graphql(client: IGraphQLClient) -> DgApiScheduleList:
    """List all schedules across all repositories."""
    result = client.execute(SCHEDULES_QUERY)

    workspace_or_error = result.get("workspaceOrError", {})
    if workspace_or_error.get("__typename") == "PythonError":
        raise Exception(f"GraphQL error: {workspace_or_error.get('message', 'Unknown error')}")

    schedules = []
    location_entries = workspace_or_error.get("locationEntries", [])

    for location_entry in location_entries:
        if location_entry.get("__typename") != "WorkspaceLocationEntry":
            continue

        location_or_load_error = location_entry.get("locationOrLoadError", {})
        if location_or_load_error.get("__typename") != "RepositoryLocation":
            continue

        repositories = location_or_load_error.get("repositories", [])

        for repo in repositories:
            repo_schedules = repo.get("schedules", [])

            for schedule_data in repo_schedules:
                schedule = DgApiSchedule(
                    id=schedule_data["id"],
                    name=schedule_data["name"],
                    cron_schedule=schedule_data["cronSchedule"],
                    pipeline_name=schedule_data["pipelineName"],
                    description=schedule_data.get("description"),
                    execution_timezone=schedule_data.get("executionTimezone", "UTC"),
                    tags=_transform_tags(schedule_data.get("tags", [])),
                    metadata_entries=_transform_metadata_entries(
                        schedule_data.get("metadataEntries", [])
                    ),
                )
                schedules.append(schedule)

    return DgApiScheduleList(items=schedules)


def get_dg_plus_api_schedule_via_graphql(
    client: IGraphQLClient, schedule_name: str
) -> DgApiSchedule:
    """Get single schedule by name.

    Note: This is simplified and assumes schedule is in the first repository found.
    For production use, we would need repository/location selection.
    """
    # First, get all schedules to find the one we want
    schedules_list = list_dg_plus_api_schedules_via_graphql(client)

    # Find the schedule by name
    for schedule in schedules_list.items:
        if schedule.name == schedule_name:
            return schedule

    raise Exception(f"Schedule not found: {schedule_name}")
