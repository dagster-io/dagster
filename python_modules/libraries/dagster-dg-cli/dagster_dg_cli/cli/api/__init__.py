"""API commands for interacting with Dagster Plus REST-like interface."""

from dagster_dg_cli.cli.api.cli_group import api_group

__all__ = ["api_group"]
