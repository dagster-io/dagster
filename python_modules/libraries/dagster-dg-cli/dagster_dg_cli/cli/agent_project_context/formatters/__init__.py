"""Output formatters for project context."""

from abc import ABC, abstractmethod

from dagster_dg_cli.cli.agent_project_context.models import ProjectContext


class BaseFormatter(ABC):
    """Base class for project context formatters."""

    @abstractmethod
    def format(self, context: ProjectContext) -> str:
        """Format project context into string output."""
        pass


def get_formatter(format_name: str) -> BaseFormatter:
    """Get formatter instance by name."""
    from dagster_dg_cli.cli.agent_project_context.formatters.json import JsonFormatter
    from dagster_dg_cli.cli.agent_project_context.formatters.markdown import MarkdownFormatter
    from dagster_dg_cli.cli.agent_project_context.formatters.xml import XmlFormatter

    formatters = {
        "json": JsonFormatter,
        "markdown": MarkdownFormatter,
        "xml": XmlFormatter,
    }

    formatter_class = formatters.get(format_name)
    if not formatter_class:
        raise ValueError(f"Unknown format: {format_name}")

    return formatter_class()
