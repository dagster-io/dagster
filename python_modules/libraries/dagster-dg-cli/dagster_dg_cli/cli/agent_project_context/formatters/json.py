"""JSON formatter for project context."""

import json

from dagster_dg_cli.cli.agent_project_context.formatters import BaseFormatter
from dagster_dg_cli.cli.agent_project_context.models import ProjectContext


class JsonFormatter(BaseFormatter):
    """Formats project context as JSON."""

    def format(self, context: ProjectContext) -> str:
        """Format project context as JSON."""
        # Convert record objects to dictionaries for JSON serialization
        context_dict = {
            "working_directory": context.working_directory,
            "project_root": context.project_root,
            "project_name": context.project_name,
            "components": [
                {
                    "key": comp.key,
                    "summary": comp.summary,
                    "namespace": comp.namespace,
                    "name": comp.name,
                }
                for comp in context.components
            ],
            "integrations": [
                {
                    "name": integ.name,
                    "description": integ.description,
                    "pypi": integ.pypi,
                }
                for integ in context.integrations
            ],
            "project_structure": {
                "has_pyproject_toml": context.project_structure.has_pyproject_toml,
                "has_dbt_project": context.project_structure.has_dbt_project,
                "dbt_project_paths": context.project_structure.dbt_project_paths,
                "python_packages": context.project_structure.python_packages,
            },
            "dbt_schemas": {
                key: {
                    "description": schema.description,
                    "scaffold_params_schema": schema.scaffold_params_schema,
                    "component_schema": schema.component_schema,
                }
                for key, schema in context.dbt_schemas.items()
            },
        }
        return json.dumps(context_dict, indent=2)
