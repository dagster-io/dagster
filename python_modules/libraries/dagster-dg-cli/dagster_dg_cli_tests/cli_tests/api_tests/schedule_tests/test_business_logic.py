"""Test schedule business logic functions without mocks.

These tests focus on testing pure functions that process data without requiring
GraphQL client mocking or external dependencies.
"""

from dagster_dg_cli.api_layer.schemas.schedule import DgApiSchedule, DgApiScheduleList
from dagster_dg_cli.cli.api.formatters import format_schedule, format_schedules


class TestFormatSchedules:
    """Test the schedule formatting functions."""

    def _create_sample_schedule_list(self):
        """Create sample DgApiScheduleList for testing."""
        schedules = [
            DgApiSchedule(
                id="schedule-1",
                name="daily_data_refresh",
                cron_schedule="0 2 * * *",
                pipeline_name="daily_data_job",
                description="Refreshes data tables daily at 2 AM",
                execution_timezone="UTC",
                tags=[
                    {"key": "team", "value": "data-platform"},
                    {"key": "priority", "value": "high"},
                ],
                metadata_entries=[
                    {"label": "owner", "description": "data-team"},
                    {
                        "label": "slack_channel",
                        "description": "#data-alerts",
                        "text": "#data-alerts",
                    },
                ],
            ),
            DgApiSchedule(
                id="schedule-2",
                name="hourly_metrics",
                cron_schedule="0 * * * *",
                pipeline_name="metrics_job",
                description="Calculates hourly metrics",
                execution_timezone="America/New_York",
                tags=[],
                metadata_entries=[],
            ),
        ]
        return DgApiScheduleList(items=schedules)

    def _create_empty_schedule_list(self):
        """Create empty DgApiScheduleList for testing."""
        return DgApiScheduleList(items=[])

    def _create_single_schedule(self):
        """Create single DgApiSchedule for testing."""
        return DgApiSchedule(
            id="schedule-1",
            name="daily_reports",
            cron_schedule="0 8 * * 1-5",
            pipeline_name="reporting_job",
            description="Generates daily reports on weekdays",
            execution_timezone="America/Los_Angeles",
            tags=[
                {"key": "environment", "value": "production"},
                {"key": "cost_center", "value": "analytics"},
            ],
            metadata_entries=[
                {
                    "label": "documentation",
                    "description": "Wiki link",
                    "url": "https://wiki.company.com/reports",
                },
                {"label": "run_timeout", "description": "Max runtime", "intValue": 3600},
                {
                    "label": "retry_policy",
                    "description": "Retry config",
                    "jsonString": '{"max_retries": 3, "delay": "30s"}',
                },
            ],
        )

    def test_format_schedules_text_output(self, snapshot):
        """Test formatting schedules as text."""
        schedule_list = self._create_sample_schedule_list()
        result = format_schedules(schedule_list, as_json=False)

        # Snapshot the entire text output
        snapshot.assert_match(result)

    def test_format_schedules_json_output(self, snapshot):
        """Test formatting schedules as JSON."""
        schedule_list = self._create_sample_schedule_list()
        result = format_schedules(schedule_list, as_json=True)

        # For JSON, we want to snapshot the parsed structure to avoid formatting differences
        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_empty_schedules_text_output(self, snapshot):
        """Test formatting empty schedule list as text."""
        schedule_list = self._create_empty_schedule_list()
        result = format_schedules(schedule_list, as_json=False)

        snapshot.assert_match(result)

    def test_format_empty_schedules_json_output(self, snapshot):
        """Test formatting empty schedule list as JSON."""
        schedule_list = self._create_empty_schedule_list()
        result = format_schedules(schedule_list, as_json=True)

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_single_schedule_text_output(self, snapshot):
        """Test formatting single schedule as text."""
        schedule = self._create_single_schedule()
        result = format_schedule(schedule, as_json=False)

        snapshot.assert_match(result)

    def test_format_single_schedule_json_output(self, snapshot):
        """Test formatting single schedule as JSON."""
        schedule = self._create_single_schedule()
        result = format_schedule(schedule, as_json=True)

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_schedule_without_metadata(self, snapshot):
        """Test formatting schedule with no metadata or tags."""
        schedule = DgApiSchedule(
            id="simple-schedule",
            name="simple_job",
            cron_schedule="0 0 * * *",
            pipeline_name="simple_pipeline",
            description=None,
            execution_timezone="UTC",
            tags=[],
            metadata_entries=[],
        )
        result = format_schedule(schedule, as_json=False)

        snapshot.assert_match(result)

    def test_format_schedule_pipeline_to_job_name_conversion(self, snapshot):
        """Test that pipeline_name is displayed as 'Job Name' in output."""
        schedule = DgApiSchedule(
            id="pipeline-test",
            name="test_schedule",
            cron_schedule="0 0 * * *",
            pipeline_name="my_pipeline_name",  # This should show as "Job Name" in output
            description="Test pipeline name conversion",
            execution_timezone="UTC",
            tags=[],
            metadata_entries=[],
        )
        result = format_schedule(schedule, as_json=False)

        # Verify "Job Name" appears instead of "Pipeline Name"
        assert "Job Name: my_pipeline_name" in result
        assert "Pipeline Name" not in result

        snapshot.assert_match(result)


class TestScheduleDataProcessing:
    """Test processing of schedule data structures.

    This class tests any pure functions related to schedule data processing
    and model creation.
    """

    def test_schedule_creation_with_complex_metadata(self, snapshot):
        """Test creating schedule with various metadata types."""
        schedule = DgApiSchedule(
            id="complex-schedule",
            name="complex_analytics_job",
            cron_schedule="0 2,14 * * *",  # Twice daily
            pipeline_name="analytics_pipeline",
            description="Complex schedule with various metadata types",
            execution_timezone="Europe/London",
            tags=[
                {"key": "team", "value": "data-engineering"},
                {"key": "sla", "value": "4h"},
                {"key": "cost_center", "value": "engineering"},
            ],
            metadata_entries=[
                {"label": "text_meta", "description": "Some text", "text": "example_text"},
                {
                    "label": "url_meta",
                    "description": "Documentation",
                    "url": "https://docs.example.com/schedules/complex",
                },
                {
                    "label": "path_meta",
                    "description": "Config file path",
                    "path": "/etc/dagster/schedules/complex.yaml",
                },
                {
                    "label": "json_meta",
                    "description": "Config",
                    "jsonString": '{"timeout": 3600, "retries": 3}',
                },
                {
                    "label": "markdown_meta",
                    "description": "Schedule notes",
                    "mdStr": "# Analytics Schedule\n\nRuns twice daily for analytics processing",
                },
                {
                    "label": "python_meta",
                    "description": "Handler class",
                    "module": "analytics.schedules",
                    "name": "ComplexScheduleHandler",
                },
                {"label": "float_meta", "description": "Success rate", "floatValue": 99.9},
                {"label": "int_meta", "description": "Average runtime", "intValue": 1800},
                {"label": "bool_meta", "description": "Auto retry enabled", "boolValue": True},
            ],
        )

        # Test JSON serialization works correctly
        result = schedule.model_dump_json(indent=2)
        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_schedule_with_complex_cron_expressions(self):
        """Test schedule creation with various cron expressions."""
        test_cases = [
            ("0 0 * * *", "daily_midnight"),
            ("0 */2 * * *", "every_two_hours"),
            ("0 9-17 * * 1-5", "business_hours_weekdays"),
            ("15 2 1 * *", "monthly_first_day"),
            ("0 0 * * SUN", "weekly_sunday"),
        ]

        for cron_expression, name in test_cases:
            schedule = DgApiSchedule(
                id=f"schedule-{name}",
                name=name,
                cron_schedule=cron_expression,
                pipeline_name=f"{name}_job",
                description=f"Schedule with cron: {cron_expression}",
                execution_timezone="UTC",
                tags=[],
                metadata_entries=[],
            )

            assert schedule.cron_schedule == cron_expression
            assert schedule.name == name

    def test_schedule_timezone_handling(self):
        """Test schedule creation with different timezones."""
        timezones = [
            "UTC",
            "America/New_York",
            "Europe/London",
            "Asia/Tokyo",
            "Australia/Sydney",
        ]

        for tz in timezones:
            schedule = DgApiSchedule(
                id=f"schedule-{tz.replace('/', '-')}",
                name=f"schedule_in_{tz.replace('/', '_')}",
                cron_schedule="0 9 * * *",
                pipeline_name="timezone_test_job",
                description=f"Schedule in {tz}",
                execution_timezone=tz,
                tags=[],
                metadata_entries=[],
            )

            assert schedule.execution_timezone == tz
