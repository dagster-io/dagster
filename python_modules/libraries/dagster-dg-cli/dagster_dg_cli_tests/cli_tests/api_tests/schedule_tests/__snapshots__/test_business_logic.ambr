# serializer version: 1
# name: TestFormatSchedules.test_format_empty_schedules_json_output[0]
  dict({
    'items': list([
    ]),
  })
# ---
# name: TestFormatSchedules.test_format_empty_schedules_text_output[0]
  ''
# ---
# name: TestFormatSchedules.test_format_schedule_pipeline_to_job_name_conversion[0]
  '''
  Name: test_schedule
  Job Name: my_pipeline_name
  Cron Schedule: 0 0 * * *
  Description: Test pipeline name conversion
  Execution Timezone: UTC
  '''
# ---
# name: TestFormatSchedules.test_format_schedule_without_metadata[0]
  '''
  Name: simple_job
  Job Name: simple_pipeline
  Cron Schedule: 0 0 * * *
  Description: None
  Execution Timezone: UTC
  '''
# ---
# name: TestFormatSchedules.test_format_schedules_json_output[0]
  dict({
    'items': list([
      dict({
        'cron_schedule': '0 2 * * *',
        'description': 'Refreshes data tables daily at 2 AM',
        'execution_timezone': 'UTC',
        'id': 'schedule-1',
        'metadata_entries': list([
          dict({
            'description': 'data-team',
            'label': 'owner',
          }),
          dict({
            'description': '#data-alerts',
            'label': 'slack_channel',
            'text': '#data-alerts',
          }),
        ]),
        'name': 'daily_data_refresh',
        'pipeline_name': 'daily_data_job',
        'tags': list([
          dict({
            'key': 'team',
            'value': 'data-platform',
          }),
          dict({
            'key': 'priority',
            'value': 'high',
          }),
        ]),
      }),
      dict({
        'cron_schedule': '0 * * * *',
        'description': 'Calculates hourly metrics',
        'execution_timezone': 'America/New_York',
        'id': 'schedule-2',
        'metadata_entries': list([
        ]),
        'name': 'hourly_metrics',
        'pipeline_name': 'metrics_job',
        'tags': list([
        ]),
      }),
    ]),
  })
# ---
# name: TestFormatSchedules.test_format_schedules_text_output[0]
  '''
  Name: daily_data_refresh
  Job Name: daily_data_job
  Cron Schedule: 0 2 * * *
  
  Name: hourly_metrics
  Job Name: metrics_job
  Cron Schedule: 0 * * * *
  '''
# ---
# name: TestFormatSchedules.test_format_single_schedule_json_output[0]
  dict({
    'cron_schedule': '0 8 * * 1-5',
    'description': 'Generates daily reports on weekdays',
    'execution_timezone': 'America/Los_Angeles',
    'id': 'schedule-1',
    'metadata_entries': list([
      dict({
        'description': 'Wiki link',
        'label': 'documentation',
        'url': 'https://wiki.company.com/reports',
      }),
      dict({
        'description': 'Max runtime',
        'intValue': 3600,
        'label': 'run_timeout',
      }),
      dict({
        'description': 'Retry config',
        'jsonString': '{"max_retries": 3, "delay": "30s"}',
        'label': 'retry_policy',
      }),
    ]),
    'name': 'daily_reports',
    'pipeline_name': 'reporting_job',
    'tags': list([
      dict({
        'key': 'environment',
        'value': 'production',
      }),
      dict({
        'key': 'cost_center',
        'value': 'analytics',
      }),
    ]),
  })
# ---
# name: TestFormatSchedules.test_format_single_schedule_text_output[0]
  '''
  Name: daily_reports
  Job Name: reporting_job
  Cron Schedule: 0 8 * * 1-5
  Description: Generates daily reports on weekdays
  Execution Timezone: America/Los_Angeles
  
  Tags:
    environment: production
    cost_center: analytics
  
  Metadata:
    documentation: https://wiki.company.com/reports
    run_timeout: 3600
    retry_policy: {"max_retries": 3, "delay": "30s"}
  '''
# ---
# name: TestScheduleDataProcessing.test_schedule_creation_with_complex_metadata[0]
  dict({
    'cron_schedule': '0 2,14 * * *',
    'description': 'Complex schedule with various metadata types',
    'execution_timezone': 'Europe/London',
    'id': 'complex-schedule',
    'metadata_entries': list([
      dict({
        'description': 'Some text',
        'label': 'text_meta',
        'text': 'example_text',
      }),
      dict({
        'description': 'Documentation',
        'label': 'url_meta',
        'url': 'https://docs.example.com/schedules/complex',
      }),
      dict({
        'description': 'Config file path',
        'label': 'path_meta',
        'path': '/etc/dagster/schedules/complex.yaml',
      }),
      dict({
        'description': 'Config',
        'jsonString': '{"timeout": 3600, "retries": 3}',
        'label': 'json_meta',
      }),
      dict({
        'description': 'Schedule notes',
        'label': 'markdown_meta',
        'mdStr': '''
          # Analytics Schedule
          
          Runs twice daily for analytics processing
        ''',
      }),
      dict({
        'description': 'Handler class',
        'label': 'python_meta',
        'module': 'analytics.schedules',
        'name': 'ComplexScheduleHandler',
      }),
      dict({
        'description': 'Success rate',
        'floatValue': 99.9,
        'label': 'float_meta',
      }),
      dict({
        'description': 'Average runtime',
        'intValue': 1800,
        'label': 'int_meta',
      }),
      dict({
        'boolValue': True,
        'description': 'Auto retry enabled',
        'label': 'bool_meta',
      }),
    ]),
    'name': 'complex_analytics_job',
    'pipeline_name': 'analytics_pipeline',
    'tags': list([
      dict({
        'key': 'team',
        'value': 'data-engineering',
      }),
      dict({
        'key': 'sla',
        'value': '4h',
      }),
      dict({
        'key': 'cost_center',
        'value': 'engineering',
      }),
    ]),
  })
# ---
